Implement EventCounts as maplike

francovs opened the pull request:
#### f044942fbcb02a01fefab9f52e102d45f7986944
<pre>
Implement EventCounts as maplike
<a href="https://bugs.webkit.org/show_bug.cgi?id=299211">https://bugs.webkit.org/show_bug.cgi?id=299211</a>
<a href="https://rdar.apple.com/160968888">rdar://160968888</a>

Reviewed by Youenn Fablet.

Implements the EventCounts interface as a maplike, which
provides methods such as .forEach(), keys(), etc.

EventCounts cannot use the previous maplike implementation
because it would get out of sync after the backing map is
initialized. To solve this, we use the trait
shouldAlwaysInitializeMapLike to decide if the backing map
should be reloaded on every access through the wrapper. The trait
simply checks for the presence of a marker
(shouldAlwaysInitializeMapLikeMarker).

This implementation isn&apos;t perfect: although the backing map
will be kept up-to-date when accessed through the wrapper,
it won&apos;t be updated if accessed through an iterator generated
from that map. This is documented by the failing test
event-timing/event-counts-maplike-iterators-always-updated.html.

Canonical link: <a href="https://commits.webkit.org/300830@main">https://commits.webkit.org/300830@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98f431c183e0cfe2190e1cb2728ae5dceaf4985d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123776 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43491 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34188 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130555 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75926 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125653 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44215 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52085 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94132 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62469 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126729 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35225 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110728 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74735 "Found 1 new API test failure: TestWTF:WTF_Condition.OneProducerOneConsumerOneSlot (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34183 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28887 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74036 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104949 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29111 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133247 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50728 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38630 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102610 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51103 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106947 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102439 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26101 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47780 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26032 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47578 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50582 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56346 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50056 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53402 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51730 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rreno commented on Source/WebCore/bindings/js/JSDOMMapLike.h:
```diff
@@ -82,6 +82,12 @@ void DOMMapAdapter::set(typename IDLKeyType::ParameterType key, typename IDLValu
 template<typename WrapperClass> JSC::JSObject& getAndInitializeBackingMap(JSC::JSGlobalObject& lexicalGlobalObject, WrapperClass& mapLike)
 {
     auto pair = getBackingMap(lexicalGlobalObject, mapLike);
+    if constexpr (requires { typename WrapperClass::DOMWrapped::reloadMapLike; }) {
```
Now that I'm looking at this more closely I don't understand why this is necessary. Don't all of the calls to Map interfaces from JS route through `getAndInitializeBackingMap`? Wouldn't you just sync the state with the native map (or whatever data structure you want to use) in `initializeMapLike`?

rreno replied:
Oh duh as soon as I ask this question I see it. In `getBackingMap` we're returning false because it doesn't think we need to initialize the object again.

youennf replied:
We could refactor the code a bit to something like:
```
if  constexpr (requires ...) {
    if (!pair.first)
         return pair.second.get();
}
DOMMapAdapter adapter...
```

rreno commented on Source/WebCore/page/EventCounts.h:
```diff
@@ -42,11 +42,10 @@ WTF_DEPRECATED_MAKE_FAST_ALLOCATED(EventCounts);
     void initializeMapLike(DOMMapAdapter&);
-    void add(EventType);
+    // Trait that causes the wrapper's backing map to be initialized again before every access:
+    using reloadMapLike = void;
```
Instead of making this live in WebCore objects why not modify the code generator to put a trait on the JS wrapper? You could invent an internal IDL attribute for this.

Something (in IDL) like

```
[ MapLikeRequiresSynchronization ] readonly maplike<DOMString, unsigned long long>
```

and then you can use template specialization for `getAndInitializeMapLike`.

```
template <typename T>
concept SynchronizingMapLike { T::MapLikeNeedsResynchronization };

template <SynchronizingMapLike WrapperClass>
JSC::JSObject& getAndInitializeBackingMap(...)
{
 ...
}
```

The names are kind of bad in the example above but the point is you can hide some of this boilerplate.

Ideally we wouldn't need to do any of this dance at all but unless someone who knows JSC runtime better than I rewrites this stuff to just use the underlying WebCore data dancing is what needs to be done.

youennf replied:
I have a slight preference for the current patch approach.
The interface WebIDL does not know itself whether the map values will change or be stable, but the C++ class knows it.
I would tend to use a different name plus a boolean instead of reloadMapLike, something like:
`constexpr static bool shouldAlwaysInitializeMapLike = true;`

rreno replied:
That makes sense. Yes a more descriptive name would help a lot for understanding what this flag is for.

francovs replied:
Will make the name more descriptive, but I'm partial to `using shouldAlwaysInitializeMapLikeTraitTag = void;` (and a matching `shouldAlwaysInitializeMapLike` trait) instead of the `constexpr static bool` approach because the latter requires checking both for the flag's existence and its value (or adding it to all objects with a maplike wrapper, or putting it in the wrapper through codegen).

youennf commented on Source/WebCore/page/EventCounts.cpp:
```diff
@@ -43,13 +43,6 @@ void EventCounts::add(EventType type)
-
 void EventCounts::initializeMapLike(DOMMapAdapter& map)
 {
     auto& eventNamesObject = eventNames();
```
It is likely overkill today, but if we find out we need to optimise, we could try to call this method only if there was a change to `m_counts` to be exposed.

francovs replied:
I wanted to add `shouldReloadMapLike()` to maplike-wrapped classes to do this, but I worry I'd be over-engineering a stopgap solution: this patch's approach is flawed with regards to iterators because iterating doesn't cause `initializeMapLike()` to be called; only creating the iterator does.

youennf commented on LayoutTests/imported/w3c/web-platform-tests/event-timing/idlharness.window-expected.txt:
```diff
@@ -25,7 +25,7 @@ PASS EventCounts interface object name
 PASS EventCounts interface: existence and properties of interface prototype object's "constructor" property
 PASS EventCounts interface: existence and properties of interface prototype object's @@unscopables property
-FAIL EventCounts interface: maplike<DOMString, unsigned long long> undefined is not an object (evaluating 'desc.value')
+PASS EventCounts interface: maplike<DOMString, unsigned long long>
```
Do we have other tests covering that feature?
For instance, is there a test that is doing the following:
- Call `entries()` a first time (maybe start iterating as well).
- Add an event (if there is no other way via internals API).
- Call entries() a second time
- Iterate on both `entries()` and validate results are as expected.

francovs replied:
Will add more testing, the event-timing tests only exercises `get()` and `.size` currently.

There is a failing test to be added:
1. Call entries and store the iterator
2. Add events
3. Iterate over the stores iterator and verify if the events added in the previous step were added

My understanding is that the stored iterator is supposed to produce results lazily, being always up-to-date. However, in this test it should access an out-of-date backing map because iterating doesn't cause `initializeMapLike()` to be called, only creating the iterator does.

youennf commented on LayoutTests/imported/w3c/web-platform-tests/event-timing/event-counts-maplike-iterators-always-updated.html:
```diff
@@ -0,0 +1,50 @@
+      }
+    })
+    assert_true(clickFound, 'click entry should exist')
+    t.done()
```
Is it needed to call t.done() here?

youennf commented on LayoutTests/imported/w3c/web-platform-tests/event-timing/event-counts-maplike-validate-iterators.html:
```diff
@@ -0,0 +1,35 @@
+    assert_true(keysIterator.next().done, "Number of entries in .keys() should match .size");
+    assert_true(valuesIterator.next().done, "Number of entries in .values() should match .size");
+    assert_true(entriesIterator.next().done, "Number of entries in .entries() should match .size");
+    t.done()
```
Ditto

youennf commented on LayoutTests/imported/w3c/web-platform-tests/event-timing/event-counts-maplike-validate-iterators.html:
```diff
@@ -0,0 +1,35 @@
+<script src=/resources/testdriver-vendor.js></script>
+<script src=resources/event-timing-test-utils.js></script>
+<script>
+  promise_test(async t => {
```
Can it be a regular test, everything seems to be synchronous here.

francovs committed:
Implement EventCounts as maplike
https://bugs.webkit.org/show_bug.cgi?id=299211
rdar://160968888

Reviewed by Youenn Fablet.

Implements the EventCounts interface as a maplike, which
provides methods such as .forEach(), keys(), etc.

EventCounts cannot use the previous maplike implementation
because it would get out of sync after the backing map is
initialized. To solve this, we use the trait
shouldAlwaysInitializeMapLike to decide if the backing map
should be reloaded on every access through the wrapper. The trait
simply checks for the presence of a marker
(shouldAlwaysInitializeMapLikeMarker).

This implementation isn't perfect: although the backing map
will be kept up-to-date when accessed through the wrapper,
it won't be updated if accessed through an iterator generated
from that map. This is documented by the failing test
event-timing/event-counts-maplike-iterators-always-updated.html.

Canonical link: https://commits.webkit.org/300830@main
